File: scripts/opscript.pl
# @(#) $Id$

sub osSubVar
{
my ($s,$vars,$params)=@_;

return $params->{lc $1}
       if $s=~/^%{?(\w+)}?$/;
return $vars->{''}->[0]->{$1}
       if $s=~/^\${?(?:\w+\.)?(\w+)}?$/;
return $vars->{lc $1}->[0]->{$2}
       if $s=~/^\${?(\w+):(?:\w+\.)?(\w+)}?$/;
return $s;
}

sub osSubVars
{
my ($s,$vars,$params)=@_;
my $c='';

while($s ne '')
     {
     my $head,$var;
     ($head,$var,$s)=$s=~m/
                           ^([^\$%]*)
			   (
			    (?:
			     (?: \${? (?:\w+:)? (?:\w+\.)? \w+ }? )
			     |
			     (?: %{? \w+ }? )
			    )?
			   )
			   (.*)$
			  /x;
     $c.=$head;
     $c.=osSubVar($var,$vars,$params) if $var ne '';
     if($head eq '' && $var eq '')
       {
       $c.=substr($s,0,1);
       substr($s,0,1)='';
       }
     }
return $c;
}

sub opScript
{
my ($dbh,$script,$params)=@_;

my $vars={};
my $result='';

my @script=();
if(ref($script))
  {
  @script=@{$script};
  }
else
  {
  @script=split "\n",$script;
  }
  
my $cont=0;
my $preamble=1;
foreach(@script)
       {
       chomp;
       if($cont)
	 {
	 $s=~s/\\\s*$/$_/;
	 }
       else
	 {
	 $s=$_;
	 }
       $cont=1,next if $s=~/\\\s*$/;
       $cont=0;
       next if $preamble && ($s=~/^\s*$/ || $s=~/^@/);
       $s=osSubVars($s,$vars,$params);
       if($s=~/^\s*\w*\s*=/)
	 {
	 my ($name,$sql)=$s=~/^\s*(\w*)\s*=(.*)$/;
	 my $sth=$dbh->prepare($sql);
	 $sth->execute;
	 $vars->{lc $name}=$sth->fetchall_arrayref({});
	 }
       else
	 {
	 $preamble=0;
	 $result.=$s."\n";
	 }
       }
return $result;
}

1;

File: lib/opscript.php
<?php
# @(#) $Id$

function osSubVar($s,$vars,$params)
{
if(preg_match('/^%{?(\w+)}?$/',$s,$match))
  return $params[strtolower($match[1])];
if(preg_match('/^\${?(?:\w+\.)?(\w+)}?$/',$s,$match))
  return $vars[''][$match[1]];
if(preg_match('/^\${?(\w+):(?:\w+\.)?(\w+)}?$/',$s,$match))
  return $vars[strtolower($match[1])][$match[2]];
return $s;
}

function osSubVars($s,$vars,$params)
{
$c='';
while($s!='')
     {
     preg_match('/^([^\$%]*)((?:(?:\${?(?:\w+:)?(?:\w+\.)?\w+}?)|(?:%{?\w+}?))?)(.*)$/',
                $s,$match);
     list($k,$head,$var,$s)=$match;
     $c.=$head;
     if($var!='')
       $c.=osSubVar($var,$vars,$params);
     if($head=='' && $var=='')
       {
       $c.=substr($s,0,1);
       $s=substr($s,1);
       }
     }
return $c;
}

function opScript($script,$params)
{
$vars=array();
$cont=0;
$preamble=1;
$out='';
foreach(explode("\n",$script) as $line)
       {
       if($cont)
         $s=preg_replace('/\\\\\s*$/',$line,$s);
       else
         $s=$line;
       if(preg_match('/\\\\\s*$/',$s))
         {
	 $cont=1;
	 continue;
	 }
       $cont=0;
       if($preamble && (preg_match('/^\s*$/',$s) || preg_match('/^@/',$s)))
         continue;
       $s=osSubVars($s,$vars,$params);
       if(preg_match('/^\s*(\w*)\s*=(.*)$/',$s,$match))
         {
	 list($k,$name,$sql)=$match;
	 $result=mysql_query($sql);
	 if(!$result)
	   die("Ошибка в операторном скрипте в операторе: $sql: ".mysql_error());
	 $vars[strtolower($name)]=mysql_fetch_assoc($result);
	 }
       else
         {
	 $preamble=0;
	 $out.="$s\n";
	 }
       }
return $out;
}
?>
